python/instrumentation/openinference-instrumentation-litellm/src/openinference/instrumentation/litellm/__init__.py

```diff
@@ -0,0 +1,182 @@
+        )
+
+    def _instrument_func_type_embedding(self, span: trace.Span, kwargs: Dict[str, Any]) -> None:
+        """
+        Currently instruments the functions:
+            litellm.embedding()
+            litellm.aembedding() (async version of embedding)
+        """
+        self._set_span_attribute(
+            span,
+            SpanAttributes.OPENINFERENCE_SPAN_KIND,
+            OpenInferenceSpanKindValues.EMBEDDING.value,
+        )
+        self._set_span_attribute(
+            span, SpanAttributes.EMBEDDING_MODEL_NAME, kwargs.get("model", "unknown_model")
+        )
+        self._set_span_attribute(span, EmbeddingAttributes.EMBEDDING_TEXT, kwargs.get("input"))
+        self._set_span_attribute(span, SpanAttributes.INPUT_VALUE, str(kwargs.get("input")))
+
+    def _instrument_func_type_image_generation(
+        self, span: trace.Span, kwargs: Dict[str, Any]
+    ) -> None:
+        """
+        Currently instruments the functions:
+            litellm.image_generation()
+            litellm.aimage_generation() (async version of image_generation)
+        """
+        self._set_span_attribute(
+            span, SpanAttributes.OPENINFERENCE_SPAN_KIND, OpenInferenceSpanKindValues.LLM.value
+        )
+        self._set_span_attribute(span, SpanAttributes.LLM_MODEL_NAME, kwargs.get("model"))
+        self._set_span_attribute(span, SpanAttributes.INPUT_VALUE, str(kwargs.get("prompt")))
+
+    def _finalize_span(self, span: trace.Span, result: Any) -> None:
+        if isinstance(result, litellm.ModelResponse):
+            self._set_span_attribute(
+                span, SpanAttributes.OUTPUT_VALUE, result.choices[0].message.content
+            )
+        elif isinstance(result, litellm.EmbeddingResponse):
+            if len(result.data) > 0:
+                first_embedding = result.data[0]
+                self._set_span_attribute(
+                    span,
+                    EmbeddingAttributes.EMBEDDING_VECTOR,
+                    json.dumps(first_embedding.get("embedding", [])),
+                )
+        elif isinstance(result, litellm.ImageResponse):
+            if len(result.data) > 0:
+                self._set_span_attribute(span, ImageAttributes.IMAGE_URL, result.data[0]["url"])
+                self._set_span_attribute(span, SpanAttributes.OUTPUT_VALUE, result.data[0]["url"])
+        if hasattr(result, "usage"):
+            self._set_span_attribute(
+                span, SpanAttributes.LLM_TOKEN_COUNT_PROMPT, result.usage["prompt_tokens"]
+            )
+            self._set_span_attribute(
+                span, SpanAttributes.LLM_TOKEN_COUNT_COMPLETION, result.usage["completion_tokens"]
+            )
+            self._set_span_attribute(
+                span, SpanAttributes.LLM_TOKEN_COUNT_TOTAL, result.usage["total_tokens"]
+            )
+
+    def _instrument(self, tracer_provider: Optional[TracerProvider] = None) -> None:
```

Comment:
To my eye it seems like the instrumentation functions are actually stateless? If so you can define them outside of the class and have this map function names to wrapping functions directly so we don't need to use something like `hasattr` and `getattr`

Reply:
Done, moved the helper functions that set span attributes to be outside of the class since they are stateless